Error occurred.

```
panic: not an App Engine context

goroutine 10 [running]:
google.golang.org/appengine/internal.fullyQualifiedAppID(0x194c308, 0xc4201d0570, 0x17eb520, 0x0)
	/Users/kazu/src/google.golang.org/appengine/internal/identity_classic.go:54 +0x95
```

Probably grpc does not support v1.7 or higher.

Please let me know if there is a way to work correctly.

--- a/gaegrpc.go
+++ b/gaegrpc.go
@@ -1,167 +1,28 @@
 package gaegrpc
 
 import (
-	"fmt"
 	"net/http"
-	"sync"
 
-	"golang.org/x/net/context"
 	"google.golang.org/appengine"
 	"google.golang.org/grpc"
-	"google.golang.org/grpc/metadata"
 )
-
-// HeaderKey is http.Request ID
-const HeaderKey = "x-gae-grpc-id"
-
-type requestKey struct{}
-
-type (
-	// ServerOption sets options such as Interceptor.
-	ServerOption func(*options)
-
-	options struct {
-		unaryInt  grpc.UnaryServerInterceptor
-		streamInt grpc.StreamServerInterceptor
-		grpcOpts  []grpc.ServerOption
-	}
-)
-
-// UnaryInterceptor returns a ServerOption
-func UnaryInterceptor(i grpc.UnaryServerInterceptor) ServerOption {
-	return func(o *options) {
-		if o.unaryInt != nil {
-			panic("The unary server interceptor was already set and may not be reset.")
-		}
-		o.unaryInt = i
-		return
-	}
-}
-
-// StreamInterceptor returns a ServerOption
-func StreamInterceptor(i grpc.StreamServerInterceptor) ServerOption {
-	return func(o *options) {
-		if o.streamInt != nil {
-			panic("The stream server interceptor was already set and may not be reset.")
-		}
-		o.streamInt = i
-		return
-	}
-}
-
-// GRPCOptions returns a ServerOption
-func GRPCOptions(opts ...grpc.ServerOption) ServerOption {
-	return func(o *options) {
-		o.grpcOpts = opts
-		return
-	}
-}
-
-var (
-	reqs = make(map[string]*http.Request)
-	mu   sync.RWMutex
-)
-
-func newContextWithRequest(ctx context.Context, r *http.Request) context.Context {
-	return context.WithValue(ctx, requestKey{}, r)
-}
-
-// requestFromContext returns *http.Request
-func requestFromContext(ctx context.Context) *http.Request {
-	if r, ok := ctx.Value(requestKey{}).(*http.Request); ok {
-		return r
-	}
-	return nil
-}
-
-func requestIDFromContext(ctx context.Context) string {
-	if md, ok := metadata.FromIncomingContext(ctx); ok {
-		v := md[HeaderKey]
-		if len(v) > 0 {
-			return v[0]
-		}
-
-	}
-	return ""
-}
-
-func newAppContext(ctx context.Context) context.Context {
-	id := requestIDFromContext(ctx)
-	if id != "" {
-		mu.RLock()
-		r := reqs[id]
-		mu.RUnlock()
-		ctx = newContextWithRequest(ctx, r)
-		ctx = appengine.WithContext(ctx, r)
-	}
-	return ctx
-}
-
-type wrapServerStream struct {
-	ctx context.Context
-	grpc.ServerStream
-}
-
-func (wss *wrapServerStream) Context() context.Context {
-	return wss.ctx
-}
-
-// requestID returns ID, ID is pointer address.
-func requestID(r *http.Request) string {
-	return fmt.Sprintf("%x", &r)
-}
 
 // NewServer returns grpc.Server for App Engine
-func NewServer(opt ...ServerOption) *grpc.Server {
-	o := &options{
-		grpcOpts: []grpc.ServerOption{},
-	}
-	for _, f := range opt {
-		f(o)
-	}
-	ops := append([]grpc.ServerOption{},
-		grpc.UnaryInterceptor(func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
-			ctx = newAppContext(ctx)
-			fmt.Printf("Test: %v", o.unaryInt)
-			if o.unaryInt != nil {
-				resp, err = o.unaryInt(ctx, req, info, handler)
-			} else {
-				resp, err = handler(ctx, req)
-			}
-			return
-		}),
-		grpc.StreamInterceptor(func(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) (err error) {
-			wss := &wrapServerStream{
-				ctx:          newAppContext(ss.Context()),
-				ServerStream: ss,
-			}
-			if o.streamInt != nil {
-				err = o.streamInt(srv, wss, info, handler)
-			} else {
-				err = handler(srv, wss)
-			}
-			return
-		}))
-	ops = append(ops, o.grpcOpts...)
-	return grpc.NewServer(ops...)
+func NewServer(opt ...grpc.ServerOption) *grpc.Server {
+	return grpc.NewServer(opt...)
 }
 
-// newRequest returns http.Request for GRPC, set the http.Request on memory
-func newRequest(r *http.Request) *http.Request {
-	id := requestID(r)
-	mu.Lock()
-	reqs[id] = r
-	mu.Unlock()
-	r.Header.Add(HeaderKey, id)
-	return r
+// NewRequest returns http.Request for GRPC, set the http.Request on memory
+func NewRequest(r *http.Request) *http.Request {
+	return r.WithContext(appengine.WithContext(r.Context(), r))
 }
 
 type wrapResponseWriter struct {
 	w http.ResponseWriter
 }
 
-// newWrapResponseWriter returns wraped http.ResponseWriter
-func newWrapResponseWriter(w http.ResponseWriter) http.ResponseWriter {
+// NewWrapResponseWriter returns wraped http.ResponseWriter
+func NewWrapResponseWriter(w http.ResponseWriter) http.ResponseWriter {
 	return &wrapResponseWriter{
 		w: w,
 	}
@@ -193,21 +54,12 @@ func (w *wrapResponseWriter) Flush() {
 	return
 }
 
-// deleteRequest deletes the http.Request on memory
-func deleteRequest(r *http.Request) {
-	mu.Lock()
-	delete(reqs, requestID(r))
-	mu.Unlock()
-}
-
 type wrapHandler struct {
 	h http.Handler
 }
 
 func (s *wrapHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	r = newRequest(r)
-	defer deleteRequest(r)
-	s.h.ServeHTTP(newWrapResponseWriter(w), r)
+	s.h.ServeHTTP(NewWrapResponseWriter(w), NewRequest(r))
 }
 
 // NewWrapHandler returns http.Handler for App Engine

--- a/gaegrpc_test.go
+++ b/gaegrpc_test.go
@@ -83,7 +83,7 @@ func (s *testService) Ping(ctx context.Context, req *testpb.PingRequest) (res *t
 }
 
 func TestPing(t *testing.T) {
-	sv := NewServer(UnaryInterceptor(func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
+	sv := NewServer(grpc.UnaryInterceptor(func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
 		id := appengine.AppID(ctx)
 		t.Logf("AppID: %v", id)
 		if want := "testapp"; id != want {

